File: epos_restaurant_2023/inventory/report/stock_transfer_report/stock_transfer_report.js

frappe.query_reports["Stock Transfer Report"] = {
	"filters": [
		{
			"fieldname":"start_date",
			"label": __("Start Date"),
			"fieldtype": "Date",
			default:frappe.datetime.get_today(),
			"reqd": 1
		},
		{
			"fieldname":"end_date",
			"label": __("End Date"),
			"fieldtype": "Date",
			default:frappe.datetime.get_today(),
			"reqd": 1
		},
		{
			"fieldname":"from_business_branch",
			"label": __("From Business Branch"),
			"fieldtype":"MultiSelectList",
			get_data: function(txt) {
				return frappe.db.get_link_options('Business Branch', txt);
			}
		},
		{
			fieldname: "from_stock_location",
			label: __("From Stock Location"),
			fieldtype: "MultiSelectList",
			get_data: function(txt) {
				return frappe.db.get_link_options('Stock Location', txt);
			}
			 
		},
		{
			"fieldname":"to_business_branch",
			"label": __("To Business Branch"),
			"fieldtype":"MultiSelectList",
			get_data: function(txt) {
				return frappe.db.get_link_options('Business Branch', txt);
			}
		},
		{
			"fieldname": "to_stock_location",
			"label": __("To Stock Location"),
			"fieldtype": "MultiSelectList",
			get_data: function(txt) {
				
				return frappe.db.get_link_options('Stock Location', txt);
			}
		},
	],

	"formatter": function(value, row, column, data, default_formatter) {
	
		value = default_formatter(value, row, column, data);

		if (data && data.is_group==1) {
			value = $(`<span>${value}</span>`);

			var $value = $(value).css("font-weight", "bold");
			

			value = $value.wrap("<p></p>").parent().html();
		}
		
		return value;
	},
};
 

File: epos_restaurant_2023/inventory/report/stock_transfer_report/stock_transfer_report.py

import frappe
from frappe import _
from py_linq import Enumerable


def execute(filters=None):
	if not filters.from_stock_location:
		filters.from_stock_location= frappe.db.get_list("Stock Location",pluck='name')
	if not filters.to_stock_location:
		filters.to_stock_location= frappe.db.get_list("Stock Location",pluck='name')
	if not filters.from_business_branch:
		filters.from_business_branch= frappe.db.get_list("Business Branch",pluck='name')
	if not filters.to_business_branch:
		filters.to_business_branch= frappe.db.get_list("Business Branch",pluck='name')
	
	data =  get_report_data(filters)
	return get_report_columns(),data,None,None, get_report_summary(data)


def get_report_data(filters):
    sql="""
    	select 
			name,
			posting_date,
			from_stock_location,
   			to_stock_location,
			from_business_branch,
			to_business_branch,
			total_quantity,
			total_amount
     	from `tabStock Transfer`
		where
			from_stock_location in %(from_stock_location)s and 
			to_stock_location in %(to_stock_location)s and 
			from_business_branch in %(from_business_branch)s and 
			to_business_branch in %(to_business_branch)s and 
			posting_date between %(start_date)s and %(end_date)s and 
			docstatus = 1

    """
    	
    data = frappe.db.sql(sql,filters , as_dict=1)
    return data

def get_report_columns():
    return [
		{"label":"Name", "fieldname":"name", "fieldtype":"Link","options":"Stock Transfer","width":150},
  		{"label":"Date", "fieldname":"posting_date", "fieldtype":"Date","align":"center","width":120},
		{"label":"From Business Branch", "fieldname":"from_business_branch","fieldtype":"Data","align":"left","width":170},
		{"label":"From Stock Location", "fieldname":"from_stock_location","fieldtype":"Data","align":"left","width":170},
		{"label":"To Business Branch", "fieldname":"to_business_branch","fieldtype":"Data","align":"left","width":150},
  		{"label":"To Stock Location", "fieldname":"to_stock_location","fieldtype":"Data","align":"left","width":150},
		{"label":"QTY", "fieldname":"total_quantity","fieldtype":"Data","align":"center","width":70},
		{"label":"Total Amount", "fieldname":"total_amount","fieldtype":"Currency","align":"right","width":110},
  		
		
	]
def get_report_summary(data,):
    report_summary = []
    report_summary.append({"label":_("QTY"),"value":frappe.utils.fmt_money(Enumerable(data).sum(lambda x: x.total_quantity or 0)),"indicator":"green"})
    report_summary.append({"label":_("Total Amount"),"value":frappe.utils.fmt_money(Enumerable(data).sum(lambda x: x.total_amount or 0)),"indicator":"red"})
    
    
    
    return report_summary